### ISSUE TITLE
YOLOv11 Segmentation Model Performance Degrades Significantly After ONNX Conversion (Missing Objects in C++ Inference)

### ISSUE THREAD

**Issue by happybear1015** [Geekgineer/YOLOs-CPP#75]
### Bug

The YOLOv11 segmentation model performs well in PyTorch (.pt) format, but its performance significantly deteriorates after conversion to ONNX. Python predictions yield accurate results (all objects in an image are segmented correctly), while C++ ONNX inference produces poor results (some objects in the same image are not segmented). Different export parameters (dynamic/half/simplify) have been tried, but the issue persists.

**Training code:**  
```
from ultralytics import YOLO  
# Load a model  
model = YOLO("yolo11m-seg.pt")    
# Train the model  
results = model.train(data="coco8-seg.yaml", epochs=100, imgsz=1280, device=[0, 1])  
```

**Python inference code:**  
```
from ultralytics import YOLO  
# Load a model  
model = YOLO("./runs/segment/train/weights/best.pt")  # pretrained YOLO11n model  

# Run batched inference on a list of images  
image1 = "datasets/PF-SPEED/images/train/ym.jpg"  
results = model([image1])  # return a list of Results objects  

# Process results list  
for result in results:  
    boxes = result.boxes  # Boxes object for bounding box outputs  
    masks = result.masks  # Masks object for segmentation masks outputs  
    keypoints = result.keypoints  # Keypoints object for pose outputs  
    probs = result.probs  # Probs object for classification outputs  
    obb = result.obb  # Oriented boxes object for OBB outputs  
    result.show()  # display to screen  
    result.save(filename="result.jpg")  # save to disk  
```

**ONNX conversion code:**  
```
from ultralytics import YOLO  
model = YOLO("./runs/segment/train/weights/best.pt")  
success = model.export(format="onnx", dynamic=True, simplify=True)
```

**C++ prediction code:**

```
#include <opencv2/highgui/highgui.hpp>
#include <iostream>
#include <string>
#include <fstream>
#include <filesystem> // For directory operations
#include <chrono>     // For high-precision timing
#include "YOLO11Seg.hpp"
#include <numeric>    // For calculating averages

namespace fs = std::filesystem; // Simplify namespace

// Get all image files in a directory
std::vector<std::string> getImageFiles(const std::string& directory) {
    std::vector<std::string> imageFiles;
    try {
        for (const auto& entry : fs::directory_iterator(directory)) {
            if (entry.is_regular_file()) {
                std::string filename = entry.path().string();
                // Check file extension
                if (filename.find(".jpg") != std::string::npos ||
                    filename.find(".png") != std::string::npos ||
                    filename.find(".jpeg") != std::string::npos) {
                    imageFiles.push_back(filename);
                }
            }
        }
    }
    catch (const fs::filesystem_error& e) {
        std::cerr << "Failed to open directory: " << directory << ", error: " << e.what() << std::endl;
    }
    return imageFiles;
}

int main()
{
    // Create results directory
    const std::string resultsDir = "results";
    if (!fs::exists(resultsDir)) {
        fs::create_directory(resultsDir);
    }

    // Paths for model, labels, and test images
    const std::string labelsPath = "../models/coco.names";
    const std::string imageDirectory = "../data"; // Image directory path
    const std::string modelPath = "../models/best.onnx";

    bool isGPU = false; // false means using CPU; true means using GPU
    // Initialize YOLO detector
    YOLOv11SegDetector detector(modelPath, labelsPath, isGPU);

    // Get all image files in directory
    std::vector<std::string> imageFiles = getImageFiles(imageDirectory);
    if (imageFiles.empty()) {
        std::cerr << "Error: No image files in directory!\n";
        return -1;
    }

    // Create timing log file
    std::ofstream timeFile(resultsDir + "/inference_times.txt");
    if (!timeFile.is_open()) {
        std::cerr << "Failed to create timing log file!\n";
        return -1;
    }

    // Store all inference times
    std::vector<long long> inferenceTimes;

    // Process each image in directory
    for (const auto& imagePath : imageFiles) {
        // Load image
        cv::Mat image = cv::imread(imagePath);
        if (image.empty()) {
            std::cerr << "Error: Failed to open or find image: " << imagePath << "\n";
            continue;
        }

        // Start pure inference timing
        auto inferenceStart = std::chrono::high_resolution_clock::now();
        std::vector<Segmentation> results = detector.segment(image, 0.0f, 0.05f);
        auto inferenceEnd = std::chrono::high_resolution_clock::now();
        auto inferenceTime = std::chrono::duration_cast<std::chrono::milliseconds>(inferenceEnd - inferenceStart);

        // Record inference time
        inferenceTimes.push_back(inferenceTime.count());
        timeFile << "Image: " << fs::path(imagePath).filename().string()
            << " Inference time: " << inferenceTime.count() << " ms\n";

        std::cout << "Instance segmentation completed, pure inference time: " << inferenceTime.count() << " ms" << std::endl;
        std::cout << "Number of detected objects: " << results.size() << std::endl;

        // Get image name (without path and extension)
        std::string imageName = fs::path(imagePath).stem().string();

        // Process each detection result (this part is not timed)
        for (size_t i = 0; i < results.size(); ++i) {
            const auto& seg = results[i];
            std::cout << "Class ID: " << seg.classId << ", Confidence: " << seg.conf << std::endl;
            std::cout << "Bounding box: (" << seg.box.x << ", " << seg.box.y << ")"
                << " Width: " << seg.box.width << " Height: " << seg.box.height << std::endl;

            // Condition: Only save mask coordinates for qualified objects
            bool isTargetObject = (seg.conf > 0.0); // Example condition: class is 0 and confidence > 0.5
            if (isTargetObject && !seg.mask.empty()) {
                // Extract mask point coordinates
                std::vector<cv::Point> maskPoints;
                cv::findNonZero(seg.mask, maskPoints); // Get coordinates of all non-zero pixels in mask

                // Write mask coordinates to file with image name and index
                std::string filename = resultsDir + "/mask_points_" + imageName + "_" + std::to_string(i) + ".txt";
                std::ofstream outFile(filename);
                if (outFile.is_open()) {
                    for (const auto& point : maskPoints) {
                        outFile << point.x << " " << point.y << "\n"; // Write coordinates
                    }
                    outFile.close();
                    std::cout << "Mask coordinates written to file: " << filename << std::endl;
                }
                else {
                    std::cerr << "Failed to open file: " << filename << std::endl;
                }
            }
            else {
                std::cout << "Object " << i << " doesn't meet requirements, skipping save." << std::endl;
            }
        }

        // Draw results (this part is not timed)
        detector.drawSegmentationsAndBoxes(image, results); // Draw masks and bounding boxes

        // Save image to file
        std::string resultImagePath = resultsDir + "/result_" + fs::path(imagePath).filename().string();
        cv::imwrite(resultImagePath, image);
    }

    // Calculate and record total and average time
    if (!inferenceTimes.empty()) {
        long long totalTime = std::accumulate(inferenceTimes.begin(), inferenceTimes.end(), 0LL);
        double averageTime = static_cast<double>(totalTime) / inferenceTimes.size();

        timeFile << "\nTotal inference time: " << totalTime << " ms\n";
        timeFile << "Average inference time: " << averageTime << " ms\n";

        std::cout << "\nProcessing completed!\n";
        std::cout << "Total inference time: " << totalTime << " ms\n";
        std::cout << "Average inference time: " << averageTime << " ms\n";
    }

    timeFile.close();
    return 0;
}
```

**Comment by happybear1015** [comment]
I tried the following code, and the results are good! Why the C++ infer result not good？

from ultralytics import YOLO
# Load a model
model = YOLO("./runs/segment/train/weights/best.onnx")  

image1="datasets/images/train/ym.jpg"
results = model([image1])


for result in results:
    boxes = result.boxes 
    masks = result.masks 
    keypoints = result.keypoints 
    probs = result.probs
    obb = result.obb 
    result.show() 
    result.save(filename="result.jpg")

**Comment by happybear1015** [comment]
I solved this problem！

**Comment by ahmedmsalah99** [comment]
@happybear1015 
Happy to hear that
Do you mind saying how did you solve it ?
